plugins/nlp/nlp.cpp

```diff
@@ -55,40 +45,22 @@ ECL_NLP_API bool getECLPluginDefinition(ECLPluginDefinitionBlock *pb)
 
 namespace nlp {
 
-    static IPluginContext * parentCtx = NULL;
+    IPluginContext * parentCtx = NULL;
+    static CriticalSection cs; 
     static NLPEng *nlpEng = NULL;
 
     ECL_NLP_API void setPluginContext(IPluginContext * _ctx) { parentCtx = _ctx; }
 
-        //--------------------------------------------------------------------------------
-        //                           ECL SERVICE ENTRYPOINTS
-        //--------------------------------------------------------------------------------
+    //--------------------------------------------------------------------------------
+    //                           ECL SERVICE ENTRYPOINTS
+    //--------------------------------------------------------------------------------
 
     ECL_NLP_API void ECL_NLP_CALL AnalyzeText(size32_t & tgtLen, char * & tgt, size32_t anaLen, const char * ana, size32_t txtLen, const char * txt)
     {
-        string w = queryCurrentProcessPath();
-        
-#ifdef NLP_DEBUG
-        ofstream handle;
-        handle.open(NLP_DEBUG_FILE, std::ofstream::out | std::ofstream::app);
-        handle << "[queryCurrentProcessPath: " << w << "]" << endl;
-#endif
-
-        size_t pos = w.find_last_of("/");
-        size_t pos2 = w.find_last_of("/",pos-1);
-        string parent = w.substr(0,pos2);
-        VStringBuffer workingFolder("%s/%s",parent.c_str(),"plugins/nlp/nlp-engine");
-
-#ifdef NLP_DEBUG
-        handle << "[parent: " << parent << "]" << endl;
-        handle << "[workingFolder nlp: " << workingFolder.str() << "]" << endl;
-        handle.close();
-#endif
-
+        CriticalBlock block(cs);
```

Comment:
@dehilsterlexis This crticial block has too large a scope.

In the example I posted I pointed out that it needed to only cover the creation of the object, but not include the call to nlpEngAnalyze, otherwise only one thread will be able to executed at once.  The same is true for the other critical section you added.

For example:
```
NLPEng * activeEngine;
{
    CriticalBlock block(cs);
    if (!nlpEng)
        nlpEng = new NLPEng();
    activeEngine = nlpEng;
}
// use activeEngine in subsequent code.
```

Reply:
I thought that is what you were meaning but I looked at other plugins and didn't see that type of use. So I was following those examples.